### PR TITLE
Router: prune channel that our peer does not have

### PR DESCRIPTION
**Pull request by sstone**
When we receive a reply to a channel query, we check whether channels that we have
but they don't are stale and can be pruned.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 1]
```diff
@@ -336,12 +328,52 @@ class Router(nodeParams: NodeParams, watcher: ActorRef) extends FSM[State, Data]
...
+
+      // don't ask for stale channel ids
+      val ourShortChannelIdsNotStale = ourShortChannelIds.filterNot(isStale(_, d.channels, d.updates))
+      val missing: SortedSet[ShortChannelId] = theirShortChannelIds -- ourShortChannelIdsNotStale
+      log.info("we received their reply, we're missing {} channel announcements/updates, format={} useGzip={}", missing.size, format, useGzip)
       val blocks = ChannelRangeQueries.encodeShortChannelIds(firstBlockNum, numberOfBlocks, missing, format, useGzip)
       blocks.foreach(block => sender ! QueryShortChannelIds(chainHash, block.shortChannelIds))
-      stay
+
+      // we have channel announcement that they don't have: check if can can prune them
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: typo: "can can"

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 2]
```diff
@@ -336,12 +328,52 @@ class Router(nodeParams: NodeParams, watcher: ActorRef) extends FSM[State, Data]
...
+      val pruningCandidates = {
+        val shortChannelIds = ourShortChannelIds -- theirShortChannelIds
+        log.info("we have {} channel that they do not have", shortChannelIds.size)
+        d.channels.filterKeys(id => shortChannelIds.contains(id))
+      }
+
+      // we limit the maximum number of channels that we will prune in one go to avoid "freezing" the app
+      val staleChannels = getStaleChannels(pruningCandidates.values, d.updates).take(MAX_PRUNE_COUNT)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: maybe apply the `.take(MAX_PRUNE_COUNT)` to `pruningCandidates.values`?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 3]
```diff
@@ -336,12 +328,52 @@ class Router(nodeParams: NodeParams, watcher: ActorRef) extends FSM[State, Data]
...
       sender ! TransportHandler.ReadAck(routingMessage)
       val (format, theirShortChannelIds, useGzip) = ChannelRangeQueries.decodeShortChannelIds(data)
-      val ourShortChannelIds: SortedSet[ShortChannelId] = d.channels.keySet.filter(keep(firstBlockNum, numberOfBlocks, _, d.channels, d.updates))
-      val missing: SortedSet[ShortChannelId] = theirShortChannelIds -- ourShortChannelIds
-      log.info("{} sent their reply fr, we're missing {} channel announcements/updates, format={} useGzip={}", remoteNodeId, missing.size, format, useGzip)
+
+      // keep our channel ids that are in [firstBlockNum, firstBlockNum + numberOfBlocks]
+      val ourShortChannelIds: SortedSet[ShortChannelId] = d.channels.keySet.filter(isInRange(_, firstBlockNum, numberOfBlocks))
+
+      // don't ask for stale channel ids
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: the comment doesn't seem to be accurate

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 4]
```diff
@@ -336,12 +328,52 @@ class Router(nodeParams: NodeParams, watcher: ActorRef) extends FSM[State, Data]
     case Event(PeerRoutingMessage(remoteNodeId, routingMessage@ReplyChannelRange(chainHash, firstBlockNum, numberOfBlocks, _, data)), d) =>
       sender ! TransportHandler.ReadAck(routingMessage)
       val (format, theirShortChannelIds, useGzip) = ChannelRangeQueries.decodeShortChannelIds(data)
-      val ourShortChannelIds: SortedSet[ShortChannelId] = d.channels.keySet.filter(keep(firstBlockNum, numberOfBlocks, _, d.channels, d.updates))
-      val missing: SortedSet[ShortChannelId] = theirShortChannelIds -- ourShortChannelIds
-      log.info("{} sent their reply fr, we're missing {} channel announcements/updates, format={} useGzip={}", remoteNodeId, missing.size, format, useGzip)
+
+      // keep our channel ids that are in [firstBlockNum, firstBlockNum + numberOfBlocks]
+      val ourShortChannelIds: SortedSet[ShortChannelId] = d.channels.keySet.filter(isInRange(_, firstBlockNum, numberOfBlocks))
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: `.channels.keySet.filter(isInRange)` doesn't seem to take into account the fact that `d.channels` is *sorted*

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 5]
```diff
@@ -526,6 +561,21 @@ object Router {
...
   }
 
+  /**
+    *
+    * @param shortChannelId short channel id
+    * @param channels channel announcements, indexed by short channel id
+    * @param updates channel updatesm indexed by channel descriptor
+    * @return true if shortChannelId is stale
+    */
+  def isStale(shortChannelId: ShortChannelId, channels: Map[ShortChannelId, ChannelAnnouncement], updates: Map[ChannelDesc, ChannelUpdate]): Boolean = {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: not sure why we need this function, why doesn't the wip-routing-sync branch need it?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 6]
```diff
@@ -535,17 +585,26 @@ object Router {
...
   }
 
+  /**
+    *
+    * @param id short channel id
+    * @param firstBlockNum first block height
+    * @param numberOfBlocks number of blocks
+    * @return true if id's height is in [firstBlockNum, firstBlockNum + numberOfBlocks]
+    */
+  def isInRange(id: ShortChannelId, firstBlockNum: Long, numberOfBlocks: Long): Boolean = {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: same as above, why do we need this function here and wip-routing-sync doesn't?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 7]
```diff
@@ -338,10 +330,50 @@ class Router(nodeParams: NodeParams, watcher: ActorRef) extends FSM[State, Data]
...
       val ourShortChannelIds: SortedSet[ShortChannelId] = d.channels.keySet.filter(keep(firstBlockNum, numberOfBlocks, _, d.channels, d.updates))
       val missing: SortedSet[ShortChannelId] = theirShortChannelIds -- ourShortChannelIds
-      log.info("{} sent their reply fr, we're missing {} channel announcements/updates, format={} useGzip={}", remoteNodeId, missing.size, format, useGzip)
+      log.info("we received their reply, we're missing {} channel announcements/updates, format={} useGzip={}", missing.size, format, useGzip)
       val blocks = ChannelRangeQueries.encodeShortChannelIds(firstBlockNum, numberOfBlocks, missing, format, useGzip)
       blocks.foreach(block => sender ! QueryShortChannelIds(chainHash, block.shortChannelIds))
-      stay
+
+      // we have channel announcement that they don't have: check if we can prune them
+      val pruningCandidates = {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: maybe we should just do:

```scala
val missing: SortedSet[ShortChannelId] = theirShortChannelIds -- ourShortChannelIds
...
val maybeStale: SortedSet[ShortChannelId] = ourShortChannelIds -- theirShortChannelIds
```

instead of trying to be smart?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 8]
```diff
@@ -338,10 +330,50 @@ class Router(nodeParams: NodeParams, watcher: ActorRef) extends FSM[State, Data]
...
+      // we first check which candidates are stale, then cap the result. We could also cap the candidate list first, there
+      // would be less calls to getStaleChannels but it would be less efficient from a "pruning" p.o.v
+      val staleChannels = getStaleChannels(pruningCandidates.values, d.updates).take(MAX_PRUNE_COUNT)
+      // then we clean up the related channel updates
+      val staleUpdates = staleChannels.map(d.channels).flatMap(c => Seq(ChannelDesc(c.shortChannelId, c.nodeId1, c.nodeId2), ChannelDesc(c.shortChannelId, c.nodeId2, c.nodeId1)))
+      // finally we remove nodes that aren't tied to any channels anymore (and deduplicate them)
+      val potentialStaleNodes = staleChannels.map(d.channels).flatMap(c => Set(c.nodeId1, c.nodeId2)).toSet
+      val channels1 = d.channels -- staleChannels
+      // no need to iterate on all nodes, just on those that are affected by current pruning
+      val staleNodes = potentialStaleNodes.filterNot(nodeId => hasChannels(nodeId, channels1.values))
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: There are no nodes in android, did you left the code on purpose? If so, why did you remove the node-related part in the `TickPruneStaleChannels` handler?

File: eclair-core/src/test/scala/fr/acinq/eclair/router/PruningSpec.scala [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      // we don't send the first 10 channels, which are stale
+      val shortChannelIds1 = shortChannelIds.drop(10)
+      val reply = ReplyChannelRange(chainHash, firstBlockNum, numberOfBlocks, 1.toByte, ChannelRangeQueries.encodeShortChannelIdsSingle(shortChannelIds1, ChannelRangeQueries.ZLIB_FORMAT, false))
+      probe.send(router, PeerRoutingMessage(remoteNodeId, reply))
+      probe.expectMsgType[QueryShortChannelIds]
+
+      // router should see that it has 10 channels that we don't have, check if they're stale, and prune them
+      awaitCond({
+        probe.send(router, 'channels)
+        val channels = probe.expectMsgType[Iterable[ChannelAnnouncement]]
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/PruningSpec.scala]: I think we should be a little more precise than that in tests, and check not only the count, but the exact channel ids

Comment (sstone) [reply eclair-core/src/test/scala/fr/acinq/eclair/router/PruningSpec.scala]: Done
